src/main/kotlin/org/jitsi/nlj/transform/node/incoming/RemoteBandwidthEstimator.kt

```diff
@@ -0,0 +1,105 @@
+import org.jitsi.rtp.rtp.header_extensions.AbsSendTimeHeaderExtension
+import org.jitsi.utils.LRUCache
+import org.jitsi.utils.logging.DiagnosticContext
+import org.jitsi.utils.logging2.Logger
+import org.jitsi_modified.impl.neomedia.rtp.remotebitrateestimator.RemoteBitrateEstimatorAbsSendTime
+import java.time.Clock
+
+/**
+ * Estimates the available bandwidth for the incoming stream using the abs-send-time extension.
+ */
+class RemoteBandwidthEstimator(
+    streamInformationStore: ReadOnlyStreamInformationStore,
+    parentLogger: Logger,
+    diagnosticContext: DiagnosticContext = DiagnosticContext(),
+    private val clock: Clock = Clock.systemUTC()
+) : ObserverNode("Remote Bandwidth Estimator") {
+    private val logger = parentLogger.createChildLogger(RemoteBandwidthEstimator::class)
+    /**
+     * The remote bandwidth estimation is enabled when REMB support is signaled, but TCC is not signaled.
+     */
+    private var enabled: Boolean by observableWhenChanged(false) {
+        _, _, newValue -> logger.debug { "Setting enabled=$newValue." }
+    }
+    private var astExtId: Int? = null
+    private var rbe: RemoteBitrateEstimatorAbsSendTime = RemoteBitrateEstimatorAbsSendTime(diagnosticContext, logger)
+    private val ssrcs = LRUCache<Long, Int>(MAX_SSRCS, true)
+    private var numRembsCreated = 0
+    private var numPacketsWithoutAbsSendTime = 0
+
+    init {
+        streamInformationStore.onRtpExtensionMapping(ABS_SEND_TIME) {
+            astExtId = it
+            logger.debug { "Setting abs-send-time extension ID to $astExtId" }
+        }
+        streamInformationStore.onRtpPayloadTypesChanged {
+            enabled = streamInformationStore.supportsRemb && !streamInformationStore.supportsTcc
+        }
+    }
+
+    companion object {
+        private const val MAX_SSRCS: Int = 8
+    }
+
+    override fun observe(packetInfo: PacketInfo) {
+        if (!enabled) return
+        val rtpPacket: RtpPacket = packetInfo.packet as? RtpPacket ?: return
+
+        astExtId?.let {
+            rtpPacket.getHeaderExtension(it)?.let { ext ->
+                val sendTimeAbsSendTime = AbsSendTimeHeaderExtension.getTime(ext).toLong()
+                rbe.incomingPacketInfoAbsSendTime(
+                    clock.millis(),
+                    sendTimeAbsSendTime,
+                    packetInfo.receivedTime,
+                    rtpPacket.length)
+                ssrcs[rtpPacket.ssrc] = 0
+            }
+        } ?: numPacketsWithoutAbsSendTime++
+    }
+
+    override fun getNodeStats(): NodeStatsBlock = super.getNodeStats().apply {
+        addString("ast_ext_id", astExtId.toString())
+        addBoolean("enabled", enabled)
+        addNumber("num_rembs_created", numRembsCreated)
+        addNumber("num_packets_without_ast", numPacketsWithoutAbsSendTime)
+    }
+
+    fun createRemb(): RtcpFbRembPacket? {
+        // REMB based BWE is not configured.
+        if (!enabled or (astExtId == null)) return null
```

Comment:
Interestingly, it looks like the infix `or` on `Boolean` doesn't do short circuiting: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-boolean/or.html

Probably not a huge deal here, but given that it's not much of an advantage over `||` maybe we want to use that?